rpc/http.go

```diff
@@ -313,3 +97,94 @@ func (client *httpClient) Close() {
 func (client *httpClient) SupportedModules() (map[string]string, error) {
 	return SupportedModules(client)
 }
+
+// httpStream reads and writes HTTP requests and responses and can be used
+// as the ReadWriteCloser for the ServerCodec.
+type httpStream struct {
+	mu              sync.Mutex
+	request         *http.Request
+	firstRead     bool
+	response        http.ResponseWriter
+	responseWritten chan struct{}
+}
+
+// Read the body of the next HTTP request.
+func (rw *httpStream) Read(p []byte) (int, error) {
+	n, err := rw.request.Body.Read(p)
+	if n == 0 && err == io.EOF && !rw.firstRead { // request already read
+		<-rw.responseWritten // wait for response to be written (http is sync)
+		return 0, io.EOF     // indicate that the request has been read
+	}
+	rw.firstRead = false
+	return n, err
+}
+
+// Write the given buffer as a HTTP response.
+// Note: this will write the headers which won't allow the response to be
+// written in multiple chunks.
+func (rw *httpStream) Write(p []byte) (int, error) {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		n, err := rw.response.Write(p)
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+		return n, err
+	}
+	return 0, errors.New("response already written")
+}
+
+// Close indicates that the HTTP response has been sent.
+func (rw *httpStream) Close() error {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
+	if rw.responseWritten != nil {
+		close(rw.responseWritten)
+		rw.responseWritten = nil
+	}
+	return nil
+}
+
+// newJSONHTTPHandler creates a HTTP handler that will parse incoming JSON requests,
+// send the request to the given API provider and sends the response back to the caller.
+func newJSONHTTPHandler(srv *Server) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.ContentLength > maxHTTPRequestContentLength {
+			http.Error(w,
+				fmt.Sprintf("content length too large (%d>%d)", r.ContentLength, maxHTTPRequestContentLength),
+				http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		// turn message based HTTP into a stream that can be used by the ServerCodec
+		rw := &httpStream{
+			request:         r,
+			response:        w,
+			firstRead:       true,
+			responseWritten: make(chan struct{}),
+		}
+
+		codec := NewJSONCodec(rw)
+		w.Header().Set("content-type", "application/json")
+
+		go srv.ServeCodec(codec)
+
+		<-rw.responseWritten
+
+		return
+	}
+}
+
+// NewHTTPServer creates a new HTTP RPC server around an API provider.
+func NewHTTPServer(corsString string, srv *Server) *http.Server {
+	c := cors.New(cors.Options{
+		AllowedOrigins: strings.Split(corsString, " "),
```

Comment:
Notably the issue is when you pass an entire command to a sub environment, then bash and other's have weird rules as to how the strip or not strip quotes around strings, or when they add or do not add extra quotes, and it makes it very error prone to keep trying to figure out how to flatten env vars in such a way as they can go inside a quoted string without breaking it. In xgo I had a particularly nasty case where I needed to convert an argument content to a bash list, then use a special iterator rule to flatten that list into something that can be appended to, and then use the new appended thing as individual parameters to some command. It took me 1-2 days or mucking around with all kinds of weird constructs only because strings and quotes were interpreted in so many ways. With commas, it's trivial to append, trivial to split.